rfcs/0004-transaction.md

```diff
@@ -0,0 +1,291 @@
+
+However, it may produce massive writes on read intensive workloads when you hope to keep the read & write operations consistent, because `GetForUpdate()` simply transforms an read operation into a write operation, rewriting the key/value pair into WAL again. Also, some researches found that manual `GetForUpdate()` often introduces unexpected hard to diagnosis bugs, as described in the "2.2 Why Serializability" section of the [Serializable Snapshot Isolation in PostgreSQL](https://arxiv.org/pdf/1208.4179) paper:
+
+> Given the existence of these techniques, one might question the
+> need to provide serializable isolation in the database: shouldn’t users
+> just program their applications to handle the lower isolation level?
+> (We have often been asked this question.) Our view is that providing
+> serializability in the database is an important simplification for
+> application developers, because concurrency issues are notoriously
+> difficult to deal with. Indeed, SI anomalies have been discovered
+> in real-world applications [14]. The analysis required to identify
+> potential anomalies (or prove that none exist) is complex and is
+> likely beyond the reach of many users. In contrast, serializable
+> transactions offer simple semantics: users can treat their transactions
+> as though they were running in isolation.
+
+As above, I think it's better to finally have SSI support in the transaction feature, but thanks to the simplicity of the RocksDB approach, we could consider to implement the RocksDB approach (Snapshot Isolation + `GetForUpdate()`) first, and then implement the SSI support after it.
+
+With regard of the SSI implementation in Badger, it's a good reference for us about SSI could be implemented as an "additive feature" over the SI.
+
+We'll cover the implementation details about the SI and SSI in the later section.
+
+### Prerequisites
+
+As described above, Snapshot and WriteBatch are considered as the prerequisites for the transaction feature.
+
+In vanilla LevelDB, Snapshot gives the "C" part of the ACID, while WriteBatch gives the "A" and "D" parts of the ACID.
+
+As a fork of LevelDB, all the transaction feature in RocksDB needs to do is to provide the Isolation part. We could regard Transaction as a pure "additive feature" over Snapshot and WriteBatch. In a too simplified view for the optimistic transaction implementation in RocksDB, the transaction feature could be regarded as:
+
+1. Adding a **conflict check before writing a WriteBatch** to guarantee the transaction isolation.
+2. Adding an **iterator over WriteBatch & Snapshot** to make MVCC possible.
```

Comment:
Probably a silly question, but what's the connection between supporting an iterator and MVCC? Is this just a way of saying we need to support read APIs on `Snapshot` and `WriteBatch`?

Reply:
in a transaction, an reader will read the WriteBatch first, if the key not exists in the write batch, it reads the Snapshot. this can be implemented as a merged iterator which automatically handles this fallback.

there might have multiple transactions writing & reading their own `WriteBatch`, and a single row might exists in diferent write batches. in my understanding, this is equivant with the MVCC mechanism of innodb's MVCC which based on UNDO logs, but what we have here is an in-memory state (WriteBatch) instead of on-disk UNDO-logs.

Reply:
Yep, this all makes sense - that's basically what I meant by supporting read apis from `WriteBatch`/`Snapshot`. Thanks!